Facilitate hash updates for GenesisBlock after hashing method update. Part of Issue 1331.

hewison-chris opened the pull request:
These commits are just to assist when we update the hashing method and the GenesisBlock hashes need to be updated.

hewison-chris committed:
Unit test: Do not rely on the transaction order in test GenesisBlock

hewison-chris committed:
Unit test: assert txs are sorted in Coinnet GenesisBlock

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -489,6 +489,50 @@ public class EnrollmentManager
         return makeEnrollment(kp, utxo, cycle_length, cache[$ - offset - 1], offset);
     }
 
+    version (unittest)
+    private static void checkGenesisEnrollments(const Block genesisBlock, KeyPair[] keys, uint cycle_length)
```
Maybe move to `genesis/package.d` along with `makeGenesis`, since they seem to be aimed at the same purpose ?
Also, style (space before parenthesis)

hewison-chris replied:
Moved and corrected.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -489,6 +489,50 @@ public class EnrollmentManager
+                keys[en.index].address)).array;
+        auto sorted_enrollments = enrollments.sort!((a,b) => a.enrol.utxo_key < b.enrol.utxo_key).array;
+        assert(genesisBlock.header.enrollments == sorted_enrollments.map!(e => e.enrol).array,
+            format!"%s\n\t],\n"(
+                sorted_enrollments
+                    .fold!((s, e) =>
+                        format!"%s\n%s"
+                        (s, format!"\t// %s\n\tEnrollment(\n\t\tHash(`%s`),\n\t\tHash(`%s`),\n\t\t%s,\n\t\tSignature(`%s`)),"
+                            (e.key, e.enrol.utxo_key, e.enrol.random_seed, e.enrol.cycle_length, e.enrol.enroll_sig)))
+                        ("\n\tenrollments: [")));
```
No tab please

hewison-chris replied:
Done

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -489,6 +489,50 @@ public class EnrollmentManager
+            Enrollment enrol;
+            PublicKey key;
+        }
+        Hash txhash = hashFull(genesisBlock.txs.filter!(tx => tx.type == TxType.Freeze).front);
+        Hash[] utxos = iota(6).map!(i => UTXO.getHash(txhash, i)).array;
```
To make slightly less assumptions, you could have an `enrollable` function similar to https://github.com/bpfkorea/agora/blob/e13aa77a2794df35d724344852960e02953b8c9d/source/agora/utils/Test.d#L646-L652

But with UTXOs instead of `TxBuilder`

hewison-chris replied:
It is not really `enrollable` as these helpers are just returning either the frozen or payment tx from Genesis. I could improve on the naming though.

hewison-chris replied:
I'll add `frozens` and `payments` as block functions.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -489,6 +489,50 @@ public class EnrollmentManager
+                            (e.key, e.enrol.utxo_key, e.enrol.random_seed, e.enrol.cycle_length, e.enrol.enroll_sig)))
+                        ("\n\tenrollments: [")));
+    }
+    /// Check the Test Genesis Block enrollments (prints replacement enrollments if needed for agora.consensus.data.genesis.Test.d)
+    unittest
+    {
+        import agora.consensus.data.genesis.Test : GenesisBlock;
+        checkGenesisEnrollments(GenesisBlock, genesis_validator_keys, 20);
+    }
+
+    /// Check the Coinnet Genesis Block enrollments (prints replacement enrollments if needed for agora.consensus.data.genesis.Coinnet.d)
+    /// This will not be used for the final Coinnet GenesisBlock which will use unknown key secrets. But can be useful for now.
+    unittest
+    {
+        import agora.consensus.data.genesis.Coinnet : GenesisBlock;
+        checkGenesisEnrollments(GenesisBlock, genesis_validator_keys, 1008);
+    }
+
```
Likewise this doesn't really belong to the EnrollmentManager.

hewison-chris replied:
Moved.

Geod24 commented on source/agora/consensus/data/genesis/Coinnet.d:
```diff
@@ -42,13 +42,18 @@ public immutable Block GenesisBlock =
     assert(GenesisBlock.header.prev_block == Hash.init);
-    assert(GenesisBlock.header.height == 0);
-    assert(GenesisBlock.header.merkle_root == GenesisBlock.merkle_tree[2]);
+    assert(GenesisBlock.header.height.value == 0);
```
This change isn't necessary, is it ?

hewison-chris replied:
Reverted.

Geod24 commented on source/agora/consensus/data/genesis/Test.d:
```diff
@@ -128,6 +128,14 @@ public immutable Block GenesisBlock = {
     ],
 };
 
+version (unittest)
+{
+    import std.algorithm;
+
+    static immutable testGenesisBlockFreezeTransaction = GenesisBlock.txs.filter!(tx => tx.type == TxType.Freeze).front;
+    static immutable testGenesisBlockPaymentTransaction = GenesisBlock.txs.filter!(tx => tx.type == TxType.Payment).front;
+}
+
```
Let's not repeat the module name in the symbol. GenesisFreezeTransaction (Tx?) should do, shouldn't it ?

hewison-chris replied:
Moved to block and renamed.

hewison-chris committed:
Unit test: assert values and sorting of enrollments in Test GenesisBlock
The enrollments hardcoded into the test and coinnet GenesisBlocks may
need to be updated. If for example the hashing method,
the validator keys or the enrollment cycle length are modified.

hewison-chris committed:
Unit test: assert txs are sorted in Test GenesisBlock

hewison-chris commented:
Tests should pass now.

AndrejMitrovic commented on source/agora/consensus/data/genesis/Coinnet.d:
```diff
@@ -60,42 +60,42 @@ unittest
 /// The enrollments that are part of the genesis block
```
What is the condition that it's asserting though? assert they're sorted? or something else? It should be in the commit message & function doc header.

AndrejMitrovic commented on source/agora/consensus/data/Block.d:
```diff
@@ -407,6 +407,16 @@ public struct Block
         return index.empty ? this.txs.length : index.front[0];
     }
+
+    public auto frozens () const @safe pure nothrow
```
```suggestion
    /// Returns a range of any freeze transactions in this block
    public auto frozens () const @safe pure nothrow
```

AndrejMitrovic commented on source/agora/consensus/data/Block.d:
```diff
@@ -407,6 +407,16 @@ public struct Block
+        return this.txs.filter!(tx => tx.type == TxType.Freeze);
+    }
+
+    public auto payments () const @safe pure nothrow
```
```suggestion
    /// Returns a range of any payment transactions in this block
    public auto payments () const @safe pure nothrow
```

AndrejMitrovic reviewed (commented):
Overall LGTM.

Geod24 commented:
@hewison-chris : Side note, remember to use an empty line between commit title and message.
E.g. this:
```
    Unit test: assert values and sorting of enrollments in Test GenesisBlock

    The enrollments hardcoded into the test and coinnet GenesisBlocks may
    need to be updated. If for example the hashing method,
    the validator keys or the enrollment cycle length are modified.
```

Not (like in this PR):
```
    Unit test: assert values and sorting of enrollments in Test GenesisBlock
    The enrollments hardcoded into the test and coinnet GenesisBlocks may
    need to be updated. If for example the hashing method,
    the validator keys or the enrollment cycle length are modified.
```

Otherwise it breaks `git log --oneline`:
<img width="1680" alt="Screen Shot 2021-02-17 at 18 28 50" src="https://user-images.githubusercontent.com/2180215/108183996-09c18f80-714e-11eb-8406-69493857df81.png">
